src/main/java/gregtech/common/covers/ender/CoverAbstractEnderLink.java

```diff
@@ -0,0 +1,410 @@
+    }
+
+    protected abstract class EntrySelectorSH extends PanelSyncHandler {
+
+        private final EntryTypes<T> type;
+        private final ModularPanel mainPanel;
+
+        protected EntrySelectorSH(ModularPanel mainPanel, EntryTypes<T> type) {
+            super(mainPanel, EntrySelectorSH::defaultPanel);
+            this.type = type;
+            this.mainPanel = mainPanel;
+        }
+
+        private static ModularPanel defaultPanel(PanelSyncManager syncManager, PanelSyncHandler syncHandler) {
+            return GTGuis.createPopupPanel("entry_selector", 168, 112);
+        }
+
+        @Override
+        public ModularPanel createUI(PanelSyncManager syncManager) {
+            List<String> names = new ArrayList<>(VirtualEnderRegistry.getEntryNames(getOwner(), type));
+            return super.createUI(syncManager)
+                    .child(IKey.lang("cover.generic.ender.known_channels")
+                            .color(UI_TITLE_COLOR).asWidget()
+                            .top(6)
+                            .left(4))
+                    .child(ListWidget.builder(names, name -> createRow(name, this.mainPanel, syncManager))
+                            .background(GTGuiTextures.DISPLAY.asIcon()
+                                    .width(168 - 8)
+                                    .height(112 - 20))
+                            .paddingTop(1)
+                            .size(168 - 12, 112 - 24)
+                            .left(4)
+                            .bottom(6));
+        }
+
+        protected IWidget createRow(String name, ModularPanel mainPanel, PanelSyncManager syncManager) {
+            T entry = VirtualEnderRegistry.getEntry(getOwner(), this.type, name);
+            String key = String.format("entry#%s_description:%d", entry.getColorStr(), isPrivate ? 1 : 0);
+            PanelSyncHandler entryDescriptionSH = syncManager.panel(key, mainPanel,
+                    (syncManager1, syncHandler) -> GTGuis.createPopupPanel("entry_description", 168, 36 + 6)
+                            .child(IKey.lang("cover.generic.ender.set_description.title", entry.getColorStr())
+                                    .color(UI_TITLE_COLOR)
+                                    .asWidget()
+                                    .left(4)
+                                    .top(6))
+                            .child(new TextFieldWidget()
+                                    .setTextColor(Color.WHITE.darker(1))
+                                    .widthRel(0.95f)
+                                    .height(18)
+                                    .value(new StringSyncValue(entry::getDescription, string -> {
+                                        entry.setDescription(string);
+                                        syncHandler.closePanel();
+                                    }))
+                                    .alignX(0.5f)
+                                    .bottom(6)));
+
+            return new Row()
```

Comment:
At the end of this row of tank entries, there should be the locked status maybe, since there is room for it. Just realized that is where the delete button appears, so maybe we can add extra space, if you think it matters.


Also, unsure if this happens currently, but are all entries visible to everyone, including locked ones? We maybe could look at making locked entries visible to only the owning player. This can be a followup PR though.

Reply:
at the moment it sets the player UUID on cover attachment, but that was copied from the ender fluid cover's logic. ideally the uuid would be the player opening the cover's UI.